File: src/java/2_CreationalPatterns/3_Prototype/PrototypeExercise.java
package com.activemesa.creational.prototype.exercise;

class Point
{
  public int x, y;

  public Point(int x, int y)
  {
    this.x = x;
    this.y = y;
  }
}

class Line
{
  public Point start, end;

  public Line(Point start, Point end)
  {
    this.start = start;
    this.end = end;
  }

  public Line deepCopy()
  {
    Point newStart = new Point(start.x, start.y);
    Point newEnd = new Point(end.x, end.y);
    return new Line(newStart, newEnd);
  }
}

File: src/js/2_CreationalPatterns/3_Prototype/prototype_exercise.js
class Point
{
  constructor(x, y)
  {
    this.x = x;
    this.y = y;
  }
}

class Line
{
  constructor(start, end)
  {
    this.start = start;
    this.end = end;
  }

  deepCopy()
  {
    let newStart = new Point(this.start.x, this.start.y);
    let newEnd = new Point(this.end.x, this.end.y);
    return new Line(newStart, newEnd);
  }
}

describe('prototype', function()
{
  it('test', function()
  {
    let line1 = new Line(
      new Point(3, 3),
      new Point(10, 10)
    );

    let line2 = line1.deepCopy();
    line1.start.x = line1.end.x =
      line1.end.x = line1.end.y = 0;

    expect(line2.start.x).toEqual(3);
    expect(line2.start.y).toEqual(3);
    expect(line2.end.x).toEqual(10);
    expect(line2.end.y).toEqual(10);
  });
});